% !TEX TS-program = pdflatex
% !TEX encoding = UTF-8 Unicode

% This is a simple template for a LaTeX document using the "article" class.
% See "book", "report", "letter" for other types of document.

\documentclass[11pt]{article} % use larger type; default would be 10pt

\usepackage[utf8]{inputenc} % set input encoding (not needed with XeLaTeX)

%%% Examples of Article customizations
% These packages are optional, depending whether you want the features they provide.
% See the LaTeX Companion or other references for full information.

%%% PAGE DIMENSIONS
\usepackage{geometry} % to change the page dimensions
\geometry{a4paper} % or letterpaper (US) or a5paper or....
% \geometry{margin=2in} % for example, change the margins to 2 inches all round
% \geometry{landscape} % set up the page for landscape
%   read geometry.pdf for detailed page layout information

\usepackage{graphicx} % support the \includegraphics command and options

% \usepackage[parfill]{parskip} % Activate to begin paragraphs with an empty line rather than an indent

%%% PACKAGES
\usepackage{hyperref} % for following hyperlinks
\hypersetup{
    colorlinks,%
    citecolor=blue,%
    filecolor=blue,%
    linkcolor=blue,%
    urlcolor=blue
}
\usepackage[usenames,dvipsnames]{color}
\usepackage{multicol}
\usepackage{longtable}
\usepackage{tabu}
\usepackage{booktabs} % for much better looking tables
\usepackage{array} % for better arrays (eg matrices) in maths
\usepackage{paralist} % very flexible & customisable lists (eg. enumerate/itemize, etc.)
\usepackage{verbatim} % adds environment for commenting out blocks of text & for better verbatim
\usepackage{subfig} % make it possible to include more than one captioned figure/table in a single float
\usepackage{pifont,mdframed}
\usepackage{lettrine}

\newenvironment{warning}
  {\par\vspace{0.3cm}\begin{mdframed}[linewidth=2pt,linecolor=red]%
    \begin{list}{}{\leftmargin=1cm}
\sffamily\item[\color{red}\ding{43}]}
  {\end{list}\end{mdframed}\par\vspace{0.3cm}}
% These packages are all incorporated in the memoir class to one degree or another...

%%% HEADERS & FOOTERS
\usepackage{fancyhdr} % This should be set AFTER setting up the page geometry
\pagestyle{fancy} % options: empty , plain , fancy
\renewcommand{\headrulewidth}{0pt} % customise the layout...
\lhead{}\chead{}\rhead{}
\lfoot{}\cfoot{\thepage}\rfoot{}

%%% SECTION TITLE APPEARANCE
\usepackage{sectsty}
\allsectionsfont{\sffamily\mdseries\upshape} % (See the fntguide.pdf for font help)
% (This matches ConTeXt defaults)

%%% ToC (table of contents) APPEARANCE
\usepackage[nottoc,notlof,notlot]{tocbibind} % Put the bibliography in the ToC
\usepackage[titles,subfigure]{tocloft} % Alter the style of the Table of Contents
\renewcommand{\cftsecfont}{\rmfamily\mdseries\upshape}
\renewcommand{\cftsecpagefont}{\rmfamily\mdseries\upshape} % No bold!

%%% Source Code Appearance
\usepackage{listings}

\definecolor{gray}{rgb}{0.4,0.4,0.4}
\definecolor{lightgray}{rgb}{0.9,0.9,0.9}
\definecolor{darkblue}{rgb}{0.0,0.0,0.6}
\definecolor{cyan}{rgb}{0.0,0.6,0.6}

\lstset{
  basicstyle=\ttfamily,
  columns=fullflexible,
  showstringspaces=false,
  commentstyle=\color{gray}\upshape
  breaklines=true
}

\lstdefinestyle{BashInputStyle}{
  language=bash,
  basicstyle=\ttfamily,
  numbersep=3pt,
  columns=fullflexible,
  backgroundcolor=\color{lightgray},
  linewidth=0.9\linewidth,
  xleftmargin=0.1\linewidth,
  breaklines=true,
  breakatwhitespace=true,
  breakautoindent=true 
}

\lstdefinelanguage{XML_new}
{
  morestring=[b]",
  morestring=[s]{>}{<},
  morecomment=[s]{<?}{?>},
  stringstyle=\color{black},
  identifierstyle=\color{darkblue},
  keywordstyle=\color{cyan},
  morekeywords={seq,drid,confcnt, totcnt, spid, flid,cov, type,url, cid, drconf}% list your attributes here
}

\lstdefinelanguage{DTD}
{
morestring=[b]",
  morecomment=[s]{<?}{?>},
%morecomment=[s]{\#}{\>},
%moredirectives={\#REQUIRED},
  stringstyle=\color{darkblue},
  identifierstyle=\color{gray},
  keywordstyle=\color{cyan},
  emphstyle=\color{RoyalPurple},
emph={CDATA,PCDATA,REQUIRED, IMPLIED},
 moredelim=*[s][\color{Black}]{(}{)},
morekeywords={ELEMENT, ATTLIST, DOCTYPE,  EMPTY}% list your attributes here
}

%\mdfdefinestyle{warnboxstyle}{%
%linecolor=green!40!black,outerlinewidth=1pt,%
%frametitlerule=true,frametitlefont=\sffamily\bfseries\color{white},%
%frametitlerulewidth=1pt,frametitlerulecolor=green!40!black,%
%frametitlebackgroundcolor=green!40!black,
%backgroundcolor=green!5,
%innertopmargin=\topskip,
%roundcorner=5pt
%}
%\newmdenv[style=warnboxstyle]{exa}
%\newenvironment{warningbox}[1]
%  {\begin{exa}[frametitle=#1]}
%  {\end{exa}}
%\optionflag{g} = -g
\newcommand{\optionflag}[1]{\texttt{-#1}}
%\optionflagarg{a}{INT} = -a INT
\newcommand{\optionflagarg}[2]{\optionflag{#1}\ \texttt{#2}}
%\optionflagoptarg{a}{INT} = -a [INT]
\newcommand{\optionflagoptarg}[2]{\optionflag{#1}\ \texttt{[#2]}}
%\optionflagargnospace{a}{INT} = -aINT
\newcommand{\optionflagargnospace}[2]{\optionflag{#1}\texttt{#2}}
%\optionflagargnospace{a}{INT} = -a[INT]
\newcommand{\optionflagoptargnospace}[2]{\optionflag{#1}\texttt{[#2]}}

\newcommand{\longoptionflag}[1]{\texttt{--#1}}
\newcommand{\longoptionflagarg}[2]{\longoptionflag{#1}\ \texttt{#2}}
\newcommand{\longoptionflagoptarg}[2]{\longoptionflag{#1}\ \texttt{[#2]}}

\newcommand{\longoptionflagargnospace}[2]{\longoptionflag{#1}\texttt{=#2}}

\newcommand{\longoptionflagoptargnospace}[2]{\longoptionflag{#1}\texttt{[=#2]}}

\newcommand{\combinedoptionflag}[2]{\optionflag{#1}\ \longoptionflag{#2}}
\newcommand{
	\combinedoptionflagarg}[3]{
		\shortstack[l]{
			\optionflagarg{#1}{#3} \\ \longoptionflagarg{#2}{#3}
		}
	}
\newcommand{
	\combinedoptionflagoptarg}[3]{
		\shortstack[l]{
			\optionflagoptarg{#1}{#3} \\ \longoptionflagoptarg{#2}{#3}
		}
	}
\newcommand{
	\combinedoptionflagargnospace}[3]{
		\shortstack[l]{
			\optionflagargnospace{#1}{#3} \\ \longoptionflagargnospace{#2}{#3}
		}
	}
\newcommand{
	\combinedoptionflagoptargnospace}[3]{
		\shortstack[l]{
			\optionflagoptargnospace{#1}{#3} \\ \longoptionflagoptargnospace{#2}{#3}
		}
	}
%%% END Article customizations

%%% The "real" document content comes below...

\title{Crass:  The CRISPR assembler (v1.0.1)}
\author{Connor Skennerton and Michael Imelfort}
\date{5th September 2016} % Activate to display a given date or no date (if empty),
         % otherwise the current date is printed 

\begin{document}
\maketitle

\tableofcontents
\newpage
\section{Quick Start}
\begin{lstlisting}[style=BashInputStyle]
	$ ./autogen.sh
	$ ./configure
	$ make
	$ make install
	
	$ ./crass [options] input_file{1, n}
        $ ./crass-assembler [--velvet | --cap3] -g INT -s LIST -x CRASS_XML_FILE -i INDIR [options]
\end{lstlisting}
\section{Installation}
Crass has been developed with the GNU build system for ease of configuration on the system.  However, because of this Crass requires a UNIX operating system and has been tested on 64-bit Linux (Ubuntu).  Crass should run on Linux 64-bit architectures with gcc installed (note that other compilers have not been tested).  Crass successfully compiles with gcc 4.4 and 4.6 other versions of gcc have not been tested.    

\subsection{Pre-installed Software and Libraries}
\begin{multicols}{3}[In short:]
\begin{itemize}
\item \href{http://gcc.gnu.org/}{gcc} \textgreater= 4.4
\item \href{http://www.gnu.org/software/autoconf/}{autoconf}  \textgreater= 2.61
\item \href{http://www.gnu.org/software/automake/}{automake} \textgreater= 1.10
\item \href{http://www.gnu.org/software/libtool/}{libtool} \textgreater= 2.2
\item \href{http://xerces.apache.org/}{Xerces-c++} \textgreater= 3.1.1
\item \href{http://zlib.net/}{zlib}
\end{itemize}
\end{multicols}
\begin{warning}
 Before installation please check the version of gcc, autoconf, libtool, automake. Some linux distributions contain very old versions as default, and this will mess up the install in obscure ways. 
\end{warning}

Crass requires that \href{http://xerces.apache.org/}{Xerces-c 3.1.1} and \href{www.zlib.net}{zlib} be installed for compilation.
\subsubsection{Installing Xerces}
If possible, try and install Xerces-c through your system's package manager making sure that you install the `dev' version of the package.  On debian based linux systems this will be a package called `libxerces-c3-dev' or similar; for systems that use RPM, the package will be called `libxerces-c-devel' or similar.  However, if you do not have root access you must install the xerces source code to a directory that you have write access to and specify the correct configuration options (see ~\nameref{sec:configure}).
\begin{warning}
The binary distributions available on the Xerces-c website are broken and you should not assume that they will work on your system.  
\end{warning}

If you installed Xerces with your systems package manager then this should be fine.  However, if you installed Xerces from source and placed it in a non-standard location, you will need to add in the \texttt{-{}-with-xerces} parameter to the \texttt{configure} script with the location of your xerces package.
\subsubsection{Optional Packages and Programs}
There are a number of optional features of Crass that require additional software.  These programs are not required for compilation but add extra features that you might find useful.  It is possible to create image files of the graphs that are produced by Crass if the \href{www.graphviz.org}{Graphviz} package is installed.  By default Crass outputs a .gv file containing Graphviz source code, so even if your running Crass on a server that doesn't have it installed you can download the .gv file to another computer for visualization.  Testing for the Graphviz package occurs during configuration.  Even if you have Graphviz installed you still need to opt-in to graph rendering (see  ~\nameref{sec:configure} for the correct options) as rendering images for a large number of graphs can take alot of time.  If you enable rendering during the configuration process you can still set user options during run-time to prevent image rendering. (see~\nameref{sec:userflags}).

Crass can also be run as a wrapper for some common assembly algorithms to assemble CRISPRs into contigs.  Currently there are two assembly wrappers, one for \texttt{velvet} and \texttt{cap3}. To activate this feature you must have the executables for either of these two programs in your \texttt{\$PATH} environment variable.  If at least one of them is in your \texttt{\$PATH} then the assembly wrapper will be compiled.  

\subsection{Compiling}
On a GNU system with the prerequisites install, simply:
\begin{lstlisting}[style=BashInputStyle]
	$ ./autogen.sh
	$ ./configure
	$ make
	$ [sudo] make install
\end{lstlisting}

\subsubsection{Configure/Compiling Options}
\label{sec:configure}
If  Xerces was installed into non-default location it is necessary to define those locations when performing \lstinline$configure$.  It may also be necessary to add the installation directories into your \lstinline$LD_LIBRARY_PATH$ environmental variable.  For example if you do not have root access you could install Xerces-c in \texttt{\~/local/} then specify that location to configure: 
\begin{lstlisting}[style=BashInputStyle]
$ ./autogen.sh
$ ./configure --with-xerces=$HOME/local --prefix=$HOME/local
$ export LD_LIBRARY_PATH=$HOME/local
$ make install
[$ echo export LD_LIBRARY_PATH=$HOME/local >> ~/.bashrc] 
\end{lstlisting}

In this example the  \texttt{--with-xerces} get passed the installation prefix for those libraries, in this case \texttt{\~/local}.  It is expected that within this directory there are directories called `include' and `lib' that contain the headers and shared objects for the libraries.


    \begin{longtable}{  l    p{10cm} }
    \hline
    Option & Definition \\  \hline\hline
   \longoptionflag{enable-verbose\_logging} & This option is different from enable-debug as it changes what information is printed in the log statments.  enabling verbose logging will add the source file name, the function name and the line number for every log statement. \\ \\    
\longoptionflag{enable-rendering} &  Set this option will allow Crass to output rendered images of the final spacer graphs.  Setting this option does not garentee that Crass will produce images, it will also need to detect at least one of the executables of the Graphviz package in your PATH environmental variable. \\  \\
   \longoptionflag{disable-assembly} & Disables the assembly wrapper even if the Xerces library and suitable genome assemblers can be found.  Disabling the assembly wrapper willl change which souce files are compiled and as a result reduce the size of the executable. \\ \\
\longoptionflag{with-xerces} & Set the path where the Xerces header and library objcets can be found.  For example if the libraries are installed in \lstinline$/opt/local/lib$ and the headers are installed in \lstinline$/opt/local/include$, specify \lstinline$--with-xerces=/opt/local$\\ \\


    \hline
    \end{longtable}

\section{Running Instructions}
Crass has two separate commands for finding and for assembling CRISPRs:
 \begin{lstlisting}[style=BashInputStyle]
	$ ./crass [options] input_file
        $ ./crass-assembler [--velvet | --cap3] -g INT -s LIST -x CRASS_XML_FILE -i INDIR [options]
\end{lstlisting}
\subsection{Finding CRISPRs}
Once installed the most minimal comand is 
\begin{lstlisting}
	$  ./crass input_file
\end{lstlisting}
which leaves all of the user options at their default values.

\subsubsection{User Flags}
\label{sec:userflags}
Crass has a large munber of user options which may seem overwhelming to a new user.  Luckily most of the options are considered 'advanced' and should never have to be changed from their default values.  Of course they are in there as user options in case you want to experiment or you suspect that your sample contains a very unusual CRISPR which may require special attention. 
    \begin{longtable}{  l   p{10cm} }
    \hline
    Option & Definition \\  \hline\hline
\\
    \combinedoptionflagarg{a}{layoutAlgorithm}{STRING} &   When enable-rendering is set and you have Graphviz installed this option will become available and allow you to change the Graphviz layout engine.  The full range of layout engines is: neato, dot, fdp, sfdp, twopi, circo \\ \\
\combinedoptionflagarg{b}{numBins}{INT} &  sets the number of colour bins used in the output spacer graph for visualising the coverage of spacers in a dataset.  By default the number of bins is equal to the range of the highest and lowest coverage for a CRISPR \\ \\
\combinedoptionflagarg{c}{graphColour}{STRING} & Changes the colour range for the output spacer graph.  There are four colour scales: red-blue, blue-red, green-red-blue, red-blue-green with the default being red-blue\\ \\
\combinedoptionflagarg{d}{minDR}{INT} & The lower bound considered acceptable for the size of a direct repeat.  The default is 23bp\\ \\
\combinedoptionflagarg{D}{maxDR}{INT} & The upper bound considered acceptable for the size of a direct repeat. The default is 47bp\\ \\
\combinedoptionflag{e}{noDebugGraph} & When the DEBUG preprocessor symbol is defined this option will become available.  When set it prevents the output of any of the debugging .gv files being produced \\ \\
\combinedoptionflagarg{f}{covCutoff}{INT} & This variable sets the minimum number of spacers allowed for a putative CRISPR to be considered real and for the assembly to be attempted.  The default is 3  \\ \\
\combinedoptionflag{g}{logToScreen} & Does not produce a log file but instead prints the contents to screen.\\ \\
\combinedoptionflag{G}{showSingletons} & Set this flag if you would like to see unconnected singleton spacers in the final graph.\\ \\
\combinedoptionflag{h}{help} & Print the basic usage and version information. \\ \\
\combinedoptionflagarg{k}{kmerCount}{INT} & Sets the number of kmers that need to be shared between putative direct repeats for them to be clustered together after the find stage.  Clustered direct repeats are eventually concatenated to form a 'true' direct repeat for a CRISPR; putative repeats that cannot be clustered are removed from consideration.  Change this variable if you feel that the clustering is too stringent and is breaking appart one CRISPR into multiple types.  The default number of kmers is 6, however the value should not be set below 6 as this would not be stringent enough; a higher value would split closely related direct repeats apart\\ \\
\combinedoptionflagarg{K}{graphNodeLen}{INT} & Crass makes a graph by cutting kmers on either side of the direct repeat and then joining these together.  The length of the kmer will dictate how connected the graph will be.  A smaller number will increase the chances of new conections being formed, however it also increases the chances of false positives.  The default value is 9.\\ \\ 
\combinedoptionflagarg{l}{logLevel}{INT} & Sets the verbosity of the log file.  Under most circumstances the log level cannot go higher than 4, unless the enable-debug option is set during configuration, which will increase the maximum value to 10.  Note that above a level of 4 alot of the information will not be understandable to the user as most of these messages are specifically for us, the developers to track down bugs.  \\ \\
\combinedoptionflag{L}{longDescription} & This changes  the names of the nodes in the spacer graph to include the sequence of the spacer.  The default is to just use the spacer ID\\ \\
\combinedoptionflag{n}{minNumRepeats} & Used only for long reads, sets the minimum number of repeats that must be identified in a read for it to be considered part of a CRISPR [default: 3]\\ \\
\combinedoptionflagarg{o}{outDir}{STRING} & Sets the output directory for files produced by Crass.  The default is the current directory\\ \\
\combinedoptionflag{r}{noRendering} & When the RENDERING preprocessor symbol is defined this option will become available.  When set it prevents the generation of rendered images from the intermeadiate debugging graphs (if DEBUG preprocessor symbol is set) and the final graphs.\\ \\
\combinedoptionflagarg{s}{minSpacer}{INT} & The lower bound considered acceptable for the size of a spacer sequence. Default is 26bp.\\ \\
\combinedoptionflagarg{S}{maxSpacer}{INT} & The upper bound considered acceptable for the size of a spacer sequence. Default is 50bp.\\ \\
\combinedoptionflag{V}{version} & Preints out program version information. \\ \\
\combinedoptionflagarg{w}{windowLength}{INT} & When using the long read search algorithm, changes the window length for finding seed sequences; can be set between 6 - 9bp.  The default value is 8bp.\\ \\ 
\hline
    \end{longtable}
\subsubsection{Output From Crass}
Once the CRISPR finder is done there will be a number of output files.  
For every CRISPR type that Crass finds it will produce a sequence file in fasta format that contains reads with that CRISPR type and a .gv file containing Graphviz source code for the arrangment of spacers.  
There will also be a file called \texttt{crass.crispr} file containing information about all the found CRISPRs in an XML format (see \nameref{sec:Fileformats} for information about the .crispr specification) and a log file (if you specified for one) that contains all the information about the run. 
Note that at this stage the CRISPR has not been assembled into contigs, however the reads have been ordered and the arragnement of spacers will be known.  
You can obtain alot of information about the CRISPRs identified from the \texttt{crass.crispr} file by using a toolkit called \href{ctskennerton.github.com/crisprtools}{crisprtools} that has been developed for this purpose.  
The two main tools that get used most often are \texttt{crisprtools stat} and \texttt{crisprtools extract}.  
The \texttt{stat} command produces a tabular summary of all the CRISPRs that were identified returning stats such as the number of spacers and the number of reads that each CRISPR was found in.
The other tool is \texttt{extract}, which can be used to get the DNA sequences of the direct repeats, spacers or flankers in fasta format from the crispr file.


\subsubsection{Visualizing Graphs}
The graph file (.gv) is important for you to visualise all of the spacer arrangments in your CRISPRs.  If your input was a metagenomic sample then it is likely that there will be multiple arrangments of spacers.  Each individual section will be numbered and this numbering is important to the Crass assembler.  The graph can be visualized using a number of programs, the ones that I use are Graphviz (\url{www.graphviz.org}) and Gephi (\url{www.gephi.org}).  Both of these programs can take a .gv file as input; Graphviz will produce a single static image whereas Gephi creates an interactive graph that can be manipulated.

\subsection{Assembling CRISPRs}
CRISPRs from metagenomic data can/will have many spacer arrangements.  
This makes them difficult for a regular genome assembler to resolve as they look for a single route through the graph.   
The Crass assembler is a wrapper for other popular genome assemblers; currently there are wrappers for Velvet and Cap3.  
I premise is that you, the user will look at the output from Crass to determine which of the paths to take through the spacer graph, which are listed as different `contigs'. 
To perform the assembly you need to tell Crass the group number of the CRISPR from the finder stage (this will be in the file name), the .crispr file from Crass and the a comma separated list of the segments/contigs IDs for your group of interest.  
The segments are listed in the names of the spacers in the graph file.  
For example a spacer with the name `sp\_445\_17\_C74' means this is spacer with id 445, in segment (contig) 17 with a coverage of 74.

  The most minimal command for the assembler is shown below: 
 \begin{lstlisting}[style=BashInputStyle]
    $ ./crass-assembler [--velvet | --cap3] -g INT -s LIST -x CRASS_XML_FILE -i INDIR [options]
\end{lstlisting}
 If for example you run Crass of a metagenome and get 5 groups and you want to assemble contigs 1, 2, 5, 6, 10 from group 3 using velvet
 \begin{lstlisting}[style=BashInputStyle]
	$ ./crass-assembler --velvet -g 3 -x crass.crispr -s 1,2,5,6,10 -i crass_out
\end{lstlisting}

This will extract all of the reads that contain spacers from those contigs and use them as input to velvet which will do the assembly in the backgroud.  This is a `dumb' process in the sense that if you provide contig IDs for multiple pathways the assembler will still try and succeed but this will undoubtibly cause breaks. 

\section{Trubleshooting}
\label{sec:trubleshooting}
\begin{longtabu} to \textwidth {X[1 , p ]  X[1 , p ]}
\hline
Error & Solution \\ \hline \hline 
\texttt{LT\_INIT: command not found} & Seen during configure, and will cause configure to fail.  Caused by the version of libtool being less than 2.2.  You will need to install a newer version of both libtool (get newer versions of autoconf and automake while your at it) and try again.  You may also need to delete the crass directory and download a fresh copy of the source code as old versions of libtool/autoconf pollute the code with outdated install helper scripts that prevent future installations \\
\texttt{crass: error while loading shared libraries} & Occurs when you have installed a dependancy for crass (Xerces-c) in a location that is non-standard.  See the example in ~\nameref{sec:configure} for how to solve this problem. \\
\texttt{floating point exception} & This error occurs when running crass and is due to divide by zero errors.  In my experience this is a compiler specific issue, usually with older versions of gcc.  If you are using a version of gcc \textgreater= 4.4 get in contact with me, otherwise please install a newer version of gcc.\\
\hline
\end{longtabu}

\section{The CRISPR File (.crispr)}
\label{sec:Fileformats}
The .crispr file is an XML format to describe all aspects of a CRISPR loci.  The bulk of the file specification will not be discussed in this manual, however the basics will be talked about to give you some idea of what it is and how Crass uses it.
\subsection{Specification Overview}
Each .crispr file contains a number of \lstinline[language=XML_new]$<group>$ tags, each one containing data about an individual CRISPR locus.  There are three sections for each group:
\paragraph{The Data Section}
This section gives you a list of all of the direct repeats (\lstinline[language=XML_new]$<drs>$), spacers (\lstinline[language=XML_new]$<spacers>$) and flanking (\lstinline[language=XML_new]$<flankers>$) that crass found.  Spacers will also have a coverage associated with them.
\paragraph{The Metadata Section}
This section lists all the other files that are associated with the CRISPR locus described in the group.  These include links to image files, and files containing raw sequence data.
\paragraph{The Assembly Section}
This represents all of the links between each spacer to all others.  Reading this section essentially equates to reading a graph of the spacer arrangement of the CRISPR locus


\section{Practical Considerations}
Crass has been designed primarily with use of Illumina data, however there is nothing stopping you from using it of data produced by other sequencing technologies such as 454, Sanger or Ion Torrent.  There are a few practical considerations when dealing with other data types.
\subsection{Read Type and Quality}
 Crass relies on exact string searching algorithms to find direct repeats in reads, which means that reads that are low quality will not produce good results.  Crass does contain a number of heuestic algorithms which try to overcome certain types of sequencing errors such as mismatches, although homopolymers require special attention (see below).  Although Crass can read fastq files it currently does not utilise quality data.  This is partly due to the inacuracy of quality scores of Illumina data, which was the original data that Crass was developed to process. 
\subsubsection{Sanger Data}
Crass does work for Sanger data and should give good results due to the long read length and high quality.  It is advised that you quality trim the data to at least Q20.  The reason is one primarily of speed of execution time as longer reads will take longer to process.
\subsubsection{Homopolymers}
Due to the underlying algorithms used in Crass, homopolymers pose a real challenge when finding direct repeats.  To try and overcome this the -{}-removeHomopolymers option tells Crass to correct these types of errors using a run-length encoding algorithm.  Basically all homopolymers are removed from the read as a preprocessing step and the 'squeezed' read is then used for direct repeat finding.  This feature is EXPERIMENTAL and at this stage does not provide optimal results.  The main issue is the rate of false positive matches is drastically increased when correcting for homopolymers so it is up to you whether you want to shift through the dud matches at the end to get out the CRISPRs that you want.  
\subsection{Bird's Nests - fixing overly connected graphs}
When running Crass some CRISPRs may end up looking like a bird's nest due to over connectidness in the graph.  This basically makes the inteptretion of the graph impossible.  You can try and resolve this by rerunning Crass on just that particular CRISPR and increasing the value of -K option.  -K sets an internal parameter when Crass builds the graph that determines how easily spacers can be joined together.  By default it is set to 11 however raising it up to a higher value may resolve the issues.  You may like to set the -G option as well to see all the extra singleton spacers that get generated.  Be warned though that in some instances Graphviz will consume significant amounts of memory when -G is set on large graphs with lots of singletons.
\subsection{Assembling}
The Crass assembler is a simple wrapper for other genome assembly algorithms.  Currently there are only two wrappers available: cap3 and velvet.  The choice of assembler should be governed by the input data so for example if Crass was processing Sanger data, it would make more sense to use cap3 as the assembler.  This isn't to say that you shouldn't experiment with different assemblers to see what's best with your data.  Finally if you want to add a new wrapper to the list just send in a bug report.
\end{document}
